### PR TITLE
Refactoring: Move sanitizer logic to their respective extensions

### PR DESCRIPTION
**Pull request by alabiaga**
Changes:
- delete sanitizer.js and move its logic to their respective extensions, bind-impl.js and amp-mustache.js
- move shared common method, rewriteAttributeValue, to url.js
- whitelist html-sanitizer usage in amp-mustache in dep-check.js
- move tests accordingly.

### CODE REVIEW

File: extensions/amp-mustache/0.1/amp-mustache.js [hunk 1]
```diff
@@ -14,16 +14,189 @@
...
-import {sanitizeHtml, sanitizeTagsForTripleMustache} from '../../../src/sanitizer';
+import {rewriteAttributeValue} from '../../../src/url';
+import {startsWith} from '../../../src/string';
+import {user} from '../../../src/log';
 
 // Configure sanitizer for output of "triple-mustache";a set of allowed tags
 // to be unescaped.
 setUnescapedSanitizier(sanitizeTagsForTripleMustache);
 
+/** @private @const {string} */
```
Comment (jridgewell) [inline extensions/amp-mustache/0.1/amp-mustache.js]: Can we move all this stuff into a separate file?

File: src/url.js [hunk 2]
```diff
@@ -527,3 +528,96 @@ export function checkCorsUrl(url) {
 export function tryDecodeUriComponent(component, opt_fallback) {
   return tryDecodeUriComponent_(component, opt_fallback);
 }
+
+/**
+ * If (tagName, attrName) is a CDN-rewritable URL attribute, returns the
```
Comment (jridgewell) [inline src/url.js]: I think these additions are too specialized for `src/url.js` (even if it already is a hodgepodge).

Comment (choumx) [reply src/url.js]: +1. The function names make sense in the context of a file like "sanitizer.js", but less so in a generic util class for URLs.

**Comment by alabiaga** [comment]
@choumx 

> I'm actually not sure this is a good idea. One thing is there's a dep-check for sanitizer.js that verifies which extensions use this functionality

I believe the only dep-check for sanitizer usage is the white listing of the amp-mustache and amp-bind extensions. As seen here:

 // Global rules
  {
    filesMatching: '**/*.js',
    mustNotDependOn: 'src/sanitizer.js',
    whitelist: [
      'extensions/amp-mustache/0.1/amp-mustache.js->src/sanitizer.js',
      'extensions/amp-ad-network-adzerk-impl/0.1/' +
          'amp-ad-network-adzerk-impl.js->src/sanitizer.js',
      'extensions/amp-bind/0.1/bind-impl.js->src/sanitizer.js',
      'extensions/amp-date-picker/0.1/amp-date-picker.js->src/sanitizer.js',
    ],
  },

The other two extensions whitelisted actually needs to be cleaned up (GH-15237), as they don't actually use it. So with my proposed changes, this dep-check will no longer need to exist.

> We might even end up restoring it depending on our implementation for SVG support.

If we end up adding support for SVG, would it not still reside in amp-mustache, the only place where would be used. If we end up reusing the JS validator and generated proto rules to validate amp mustache template svg usage then it would then reside with the rest of the validator logic. I think having this sanitizer file is just confusing as its usage is very isolated.

**Comment by choumx** [comment]
Another approach could be to create a subfolder in `src/` for code that's shared by extensions but should not be used in the main binary.

**Comment by alabiaga** [comment]
@choumx Yea I am ok with this idea as well, as I now have a better understanding of how things work. 

I initially proposed this PR because I thought it would have been easier for a newcomer to understand the usage of the sanitizing logic if it just resided closer to the codebase where they were being used, as I found it confusing that this logic was under src...as I was under the impression that everything under src was shipped with the main binary and used in many places. But yes, as you mentioned this sanitizing of html or URLs can be used elsewhere in the future so it is probably best it resides in it own file. I don't feel strongly about these changes anymore, so we just throw away this PR.

**Comment by choumx** [comment]
Thanks for being flexible Aaron and sorry about the initial confusion. Our code documentation could be a lot better.

**Pull request closed without merging**